kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/HighOffsetMetricsExposer.scala

```diff
@@ -0,0 +1,121 @@
+}
+
+case class MetricMap(metricMap: Map[String, (Map[Int, (Metric.Counter[Long], Long)], Metric.Counter[Long])]) {
+
+  def updateWithTopicParitionChanges(
+      changes: KafkaTopicPartitionInfoChanges,
+      config: MetricExposerConfig
+  ): (MetricMap, List[(Metric.Counter[Long], Long)]) = {
+    val addedOrUpdated = changes.addedOrUpdated
+    val removed = changes.removed
+    var newMetricMap = metricMap
+    var metricUpdates: List[(Metric.Counter[Long], Long)] = List.empty
+    // Add new metrics and update stale ones
+    for ((topicPartition, partitionInfo) <- addedOrUpdated) {
+
+      newMetricMap = newMetricMap.updatedWith(topicPartition.topic) {
+        case Some((topicMap: Map[Int, (Metric.Counter[Long], Long)], aggregateCounter: Metric.Counter[Long])) =>
+          val newTopicMap = topicMap.updatedWith(topicPartition.partition) {
+            case Some((counter, lastValue)) =>
+              val increase = partitionInfo.endOffset - lastValue
+              metricUpdates = metricUpdates :+ (aggregateCounter, increase)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, increase)
+              }
+              Some((counter, partitionInfo.endOffset))
+            case None =>
+              val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+              metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
+              }
+              Some((counter, partitionInfo.endOffset))
+          }
+          Some((newTopicMap, aggregateCounter))
+        case None =>
+          val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+          val newTopicMap = Map.newBuilder
+            .addOne(
+              topicPartition.partition,
+              (counter, partitionInfo.endOffset)
+            )
+            .result()
+          val aggregateCounter = Helper.newTopicMetric(topicPartition.topic)
+          metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
+          if (config.perPartitionMetrics) {
+            metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
+          }
+          Some((newTopicMap, aggregateCounter))
+      }
+    }
+
+    // Update the metric map to reflect removals
+    for (topicPartition <- removed) {
+      val topic = topicPartition.topic
+      val partition = topicPartition.partition
+
+      newMetricMap = newMetricMap.updatedWith(topic) {
+        case Some((map: Map[Int, (Metric.Counter[Long], Long)], aggregateCounter: Metric.Counter[Long])) =>
+          val oldMetricValue = map.get(partition).map(_._2).getOrElse(0L)
+          val newMap = map.removed(partition)
```

Comment:
I'm not sure what happens to the per-partition metric if we just remove it from our map - does it get GC-d and removed from the prometheus export?

Reply:
As discussed, we will set the metric to be zero in the case that a partition is deleted